tensorboard/plugins/profile/tf_op_profile/tf-op-table.html

```diff
@@ -0,0 +1,209 @@
+<!--
+@license
+Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+<link rel="import" href="utils.html">
+
+<dom-module id="tf-op-table-styles"><template><style>
+#time {
+  display: inline-block;
+  width: 80px;
+}
+#name {
+  display: inline-block;
+  min-width: 40%;
+  box-sizing: border-box;
+}
+#row, #header {
+  padding: 0.5em;
+}
+</style></template></dom-module>
+
+<!--
+  tf-op-table-entry is a row in a tf-op-table.
+  It describes an op, or group of ops (the node property).
+  Its background contains a bar, whose size shows the time taken, and whose
+  color shows the op's utilization.
+  If the node has children, the tf-op-table-entry has child entries, and can be
+  expanded/collapsed.
+-->
+<dom-module id="tf-op-table-entry">
+<template>
+<style include="tf-op-table-styles">
+#row {
+  position: relative;
+  box-shadow: inset 0 1px 0 0 rgba(0,0,0,0.06);
+  z-index: 0;
+}
+#bar {
+  position: absolute;
+  top: 0;
+  left: 0;
+  bottom: 0;
+  z-index: -1;
+}
+:host {
+  display: block;
+}
+:host.selected {
+  background-color: rgba(0,0,0,0.04);
+}
+:host.selected > #row {
+  font-weight: bold;
+}
+#row:hover {
+  background-color: rgba(0,0,0,0.05);
+}
+#disclosure {
+  display: inline-block;
+  width: 1em;
+}
+#name {
+  font-family: monospace;
+}
+</style>
+<div id="row"
+     on-click="_handleHeaderClick"
+     on-mouseenter="_handleHeaderMouseEnter"
+     on-mouseleave="_handleHeaderMouseLeave"
+     hidden="[[!level]]">
+  <div id="bar"></div>
+  <span id="time">{{_percent(node.metrics.time)}}</span>
+  <span id="name" style$="padding-left:[[level]]em;">
+    <span id="disclosure">
+      <span hidden="[[!node.children.length]]">
+        <span hidden="[[expanded]]">&#x25b6;</span>
+        <span hidden="[[!expanded]]">&#x25bc;</span>
+      </span>
+    </span>{{node.name}}
+  </span>
+  <span id="provenance">{{_provenance(node)}}</span>
+</div>
+<template is="dom-if" if="[[expanded]]">
+  <template is="dom-repeat" items="{{node.children}}" sort="_sortEntry">
+    <tf-op-table-entry
+      node="[[item]]"
+      level={{_nextLevel(level)}}
+      table={{table}}>
+    </tf-op-table-entry>
+  </template>
+</template>
+</dom-module>
+
+<script>
+import {flameColor, percent, utilization} from './utils.js';
+
+Polymer({
+  is: 'tf-op-table-entry',
+  properties: {
+    node: {
+      type: Object,
+    },
+    level: {
+      type: Number,
+      value: 0,
+    },
+    table: {
```

Comment:
I see. Indeed, 10K function calls is a lot, and this harks to a downside of using binding :( It's hard to determine what logic runs when notifications trigger. Lets just use `this` then. Could you add a comment to `tf-op-table-entry` noting why we pass in the table?

Another thing you could do is make `tf-op-table` pass a callback to `tf-op-table-entry` that is called within `_handleHeaderClick ` after expansion is toggled, ie

in `tf-op-table-entry`:

```js
// A callback called when the header is clicked.
// The callback is passed this table entry.
headerClickCallback: Object,
...
_handleHeaderClick: function(e) {
  this.expanded ^= true;
  if (this.headerClickCallback) {
    this.headerClickCallback(this);
  }
  this.classList.add('selected');
},
```

and then in `tf-op-table`:

```html
<tf-op-table-entry ... headerClickCallback="[[_headerClickCallback]]"> ...
```

```js
_handleHeaderClick: {
  type: Object,
  readOnly: true,
  // This property returns a function that returns a function.
  // We actually want this property to be a function, and polymer calls
  // the higher-level function first to initialize the property.
  value: () => ((newlySelectedEntry) => {
    if (this._selectedEntry) {
      // TODO(chizeng): Define an isSelected property on the entry, and have the
      // entry toggle its own class. The table would merely set `isSelected` on the
      // entry.
      this._selectedEntry.classList.remove('selected');
    }
    this._selectedEntry = newlySelectedEntry;
    this.selected = newlySelectedEntry.node;
  }),
},
```

The advantage here is that each component is only deals with itself. `tf-op-table` should not have to set classes on `tf-op-table-entry`,  and `tf-op-table-entry` should not have to manipulate properties of `tf-op-table` directly.

That being said, it seems like you all gave good thought to the design in terms of performance, so this has my approval either way.